### ISSUE TITLE
Error: Cannot read property 'indexOf' of undefined & Could not resolve external type c: 

### ISSUE THREAD

**Issue by MrBokeh** [rossmartin/cordova-plugin-instagram-assets-picker#1]
Hi Ross,

I have tried to use it but running into a lot of warning: Could not resolve external type c: xxxxxxxx
Then I tryied to uninstall it and it shows me this error:
`
Instagram Assets Picker BeforePluginUninstall.js, attempting to modify build.xcconfig
xcConfigBuildFilePath:  /Users/xxxxxxxx/platforms/ios/cordova/build.xcconfig
Error: Cannot read property 'indexOf' of undefined
`

**Comment by rossmartin** [comment]
When you install do you see any reference to the `AfterPluginInstall.js` hook as being a culprit?  This hook adds an entry to your header search paths in the Xcode project so you don't have to manually in the project.

What cordova CLI version do you have? `cordova -v` will show.  What version of the iOS platform are you running?

I'm also interested in what node version you have, `node -v` will show.  

Are you installing the plugin from the root of your cordova project?

**Comment by MrBokeh** [comment]
cordova -v
6.0.0
node -v
v4.3.2

yes, I installed it from the root of the project, I cannot uninstall the ios plugin:
`Uninstalling cordova-plugin-instagram-assets-picker from android
Instagram Assets Picker BeforePluginUninstall.js, attempting to modify build.xcconfig
xcConfigBuildFilePath:  /Users/xxxxxx/platforms/ios/cordova/build.xcconfig
Error: Cannot read property 'indexOf' of undefined`

**Comment by MrBokeh** [comment]
the when I build the ios projectI will get a bunch of these warnings:
`
warning: /var/folders/1q/29tmthy113g05p41tbdxrg3r0000gn/C/org.llvm.clang.rossmartin/ModuleCache/QLJ9RU451UUY/Dispatch-2LX9MWM6UEPQW.pcm: No such file or directory`

**Comment by rossmartin** [comment]
I see warnings when I build via the cordova or ionic CLI with this plugin but my app runs fine.  I'm not sure exactly why your install didn't go smooth but I think it is something to do with the `AfterPluginInstall.js` hook.

Can you please try my app that uses the video editor and this plugin and let me know if you have any issues? https://github.com/rossmartin/video-editor-ionic2

I'll help get this issue fixed, thanks for letting me know.

**Comment by MrBokeh** [comment]
Sure I am going to give it a shot, actually the app that has problem was with ionic one and the ionic info is here:

```
Your system information:

Cordova CLI: 6.0.0
Ionic Version: 1.2.4-nightly-1917
Ionic CLI Version: 2.0.0-beta.19
Ionic App Lib Version: 2.0.0-beta.9
ios-deploy version: 1.8.5 
ios-sim version: 5.0.6 
OS: Mac OS X El Capitan
Node Version: v4.3.2
Xcode version: Xcode 7.2.1 Build version 7C1002 
```

**Comment by rossmartin** [comment]
I thought it might be something to do with using ES6 in the plugin hook scripts but I don't think it is that.  I think for some reason the "before plugin install" hook couldn't find your `build.xcconfig` file to add the GPUImageHeaders to the `HEADER_SEARCH_PATHS`.  I automated a lot of the steps required to use GPUImage in a project and I might not be able to add the GPUImage header files to your header search paths automagically.

**Comment by MrBokeh** [comment]
Is there a way that I can uninstall the ios plugin manually and then re-install it again?

Btw your ionic 2 app is working and it's so beautiful lol

**Comment by MrBokeh** [comment]
wait a sec, the record or import buttons doesnt work tho

**Comment by rossmartin** [comment]
Hmmm I have seen cases where the status bar plugin doesn't init and it takes building it a couple more times for it to work.  I haven't figured out if that is an issue with Ionic 2 or the status bar plugin.  Can you trying building and running a few more times to see if it works?

**Comment by rossmartin** [comment]
Also, are you running this on a device from Xcode or on the simulator?

**Comment by MrBokeh** [comment]
Running it on my iPhone 6 Plus with ios 9.2

Btw I saw your code didn't use ionic native which is interesting to see that you can call your libraries directly without any import. Also I can see that you have a lib folder on the www folder which already uglified.

**Comment by rossmartin** [comment]
When you run `ionic plugin ls` what do you see?

**Comment by MrBokeh** [comment]
No plugins added. Use `cordova plugin add <plugin>`

**Comment by rossmartin** [comment]
After you did step #4 for trying video-editor-ionic2 did you do `ionic platform add ios` ?

If so that is odd that it didn't install the plugins listed in the projects `package.json` file.

If you did do the `ionic platform add ios` after step #4 can you try adding the plugins below?

```
ionic plugin add cordova-plugin-camera cordova-plugin-device cordova-plugin-media-capture https://github.com/driftyco/ionic-plugin-keyboard.git cordova-plugin-statusbar cordova-plugin-spinner-dialog cordova-plugin-instagram-assets-picker cordova-plugin-video-editor
```

**Comment by MrBokeh** [comment]
Same issue occur now, I guess when I have premission issue when adding the plugin:
`Error occurred while adding plugin:  Error: EACCES: permission denied, scandir '/Users/xxxxxx/video-editor-ionic2/platforms/ios/build/sharedpch'`

Then the plug in added half way and then when I try to use sudo to reinstall:

```
Error: Uh oh!
"/Users/xxxxxx/video-editor-ionic2/platforms/ios/VideoEditor/Plugins/cordova-plugin-instagram-assets-picker/InstagramAssetsPicker.m" already exists!

Fetching plugin "cordova-plugin-instagram-assets-picker" via npm
Installing "cordova-plugin-instagram-assets-picker" for ios
Error during processing of action! Attempting to revert...
Failed to install 'cordova-plugin-instagram-assets-picker':Error: Uh oh!
```

It cannot revert and it stopped then when I try to uninstall the plugin with your ionic 2 app like what I did on my ionic 1 app:

```
sudo ionic plugin rm cordova-plugin-instagram-assets-picker
Instagram Assets Picker BeforePluginUninstall.js, attempting to modify build.xcconfig
xcConfigBuildFilePath:  /Users/xxxx/video-editor-ionic2/platforms/ios/cordova/build.xcconfig
Error: Cannot read property 'indexOf' of undefined
```

**Comment by MrBokeh** [comment]
Every time I update Cordova and Ionic, the permissions of two libraries are reset to none and I will need to manually go to those problem folders to give access permissions to those libraries which is so annoying. Now I get how these occurs now.

**Comment by rossmartin** [comment]
Adrian I recommend ["fixing" your npm permissions](https://docs.npmjs.com/getting-started/fixing-npm-permissions) in case there is an issue there.  Also check out this thread [here](https://forum.ionicframework.com/t/error-occurred-trying-to-fall-back-to-cordova-lib-execution/38546/10).

I'm going to update the `AfterPluginInstall` and `BeforePluginInstall` hooks on this plugin to account for a situation where it can't find your `build.xcconfig` file.

**Comment by rossmartin** [comment]
Here are the necessary steps I recommend trying - 

uninstall globally (use sudo if needed)

```
sudo npm uninstall -g ionic
sudo npm uninstall -g ionic@beta
sudo npm uninstall -g cordova
```

[Fix permissions like this](https://docs.npmjs.com/getting-started/fixing-npm-permissions24)

`sudo chown -R ${whoami}  ~/.config`

reinstall (without sudo)

```
npm install -g cordova
npm install -g ionic@beta
```

**Comment by MrBokeh** [comment]
So what I did was uninstall those and reinstall node plus these lines and the permission is now fixed:

```
sudo chown -R $(whoami) $(npm config get prefix)/{lib/node_modules,bin,share}
sudo chown -R myusername /usr/local/
sudo chown -R myusername /Users/myusername/.config
```

**Comment by rossmartin** [comment]
🍻 are you able to use the plugin and the other app now?

**Comment by MrBokeh** [comment]
```
ionic plugin add cordova-plugin-instagram-assets-picker
Fetching plugin "cordova-plugin-instagram-assets-picker" via npm
Installing "cordova-plugin-instagram-assets-picker" for android
Installing "cordova-plugin-instagram-assets-picker" for ios
Instagram Assets Picker AfterPluginInstall.js, attempting to modify build.xcconfig
xcConfigBuildFilePath:  /Users/xxxxxxx/platforms/ios/cordova/build.xcconfig
Failed to install 'cordova-plugin-instagram-assets-picker':TypeError: Cannot read property 'indexOf' of undefined
    at Object.<anonymous> (/Users/xxxxxxx/plugins/cordova-plugin-instagram-assets-picker/hooks/AfterPluginInstall.js:21:38)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at runScriptViaModuleLoader (/usr/local/lib/node_modules/cordova/node_modules/cordova-lib/src/hooks/HooksRunner.js:159:20)
    at runScript (/usr/local/lib/node_modules/cordova/node_modules/cordova-lib/src/hooks/HooksRunner.js:145:16)
    at /usr/local/lib/node_modules/cordova/node_modules/cordova-lib/src/hooks/HooksRunner.js:130:20
Error: Cannot read property 'indexOf' of undefined
```

**Comment by MrBokeh** [comment]
No it has the same issue, I rollbacked my code and try to reinstall it and I got the error above...

**Comment by MrBokeh** [comment]
Then again when I tried to uninstall it:

```
ionic plugin rm cordova-plugin-instagram-assets-picker
Uninstalling cordova-plugin-instagram-assets-picker from android
Instagram Assets Picker BeforePluginUninstall.js, attempting to modify build.xcconfig
xcConfigBuildFilePath:  /Users/xxxxxx/platforms/ios/cordova/build.xcconfig
Error: Cannot read property 'indexOf' of undefined
```

**Comment by MrBokeh** [comment]
```
ionic build ios
An Error occurred trying to fall back to Cordova-lib execution: TypeError: Cannot read property 'then' of undefined
    at Object.IonicTask.runCordova (/usr/local/lib/node_modules/ionic/lib/ionic/cordova.js:296:21)
    at /usr/local/lib/node_modules/ionic/lib/ionic/cordova.js:162:19
    at _fulfilled (/usr/local/lib/node_modules/ionic/node_modules/q/q.js:787:54)
    at self.promiseDispatch.done (/usr/local/lib/node_modules/ionic/node_modules/q/q.js:816:30)
    at Promise.promise.promiseDispatch (/usr/local/lib/node_modules/ionic/node_modules/q/q.js:749:13)
    at /usr/local/lib/node_modules/ionic/node_modules/q/q.js:557:44
    at flush (/usr/local/lib/node_modules/ionic/node_modules/q/q.js:108:17)
    at _combinedTickCallback (node.js:376:9)
    at process._tickCallback (node.js:407:11)
Error happened [TypeError: Cannot read property 'then' of undefined]
TypeError: Cannot read property 'then' of undefined
    at Object.IonicTask.runCordova (/usr/local/lib/node_modules/ionic/lib/ionic/cordova.js:296:21)
    at /usr/local/lib/node_modules/ionic/lib/ionic/cordova.js:162:19
    at _fulfilled (/usr/local/lib/node_modules/ionic/node_modules/q/q.js:787:54)
    at self.promiseDispatch.done (/usr/local/lib/node_modules/ionic/node_modules/q/q.js:816:30)
    at Promise.promise.promiseDispatch (/usr/local/lib/node_modules/ionic/node_modules/q/q.js:749:13)
    at /usr/local/lib/node_modules/ionic/node_modules/q/q.js:557:44
    at flush (/usr/local/lib/node_modules/ionic/node_modules/q/q.js:108:17)
    at _combinedTickCallback (node.js:376:9)
    at process._tickCallback (node.js:407:11)

```

**Comment by MrBokeh** [comment]
The ionic build is not working either without sudo, is that because of the Ionic Version: 1.2.4-nightly-1917 is a nightly build?

**Comment by rossmartin** [comment]
Ok you'll not want to use sudo for any ionic or npm commands otherwise it will result in the problem with `ionic build` that you see.

To fix the issue with building run -

`sudo chown -R ${whoami} ~/.config`

I'm not at home currently will have to look into this more tomorrow.

**Comment by MrBokeh** [comment]
I still have no luck to remove it, how can I do all these add or remove manually? I wanna cry.

**Comment by MrBokeh** [comment]
Does this new update work with ionic 1?

**Comment by rossmartin** [comment]
This was fixed via  599214e8d77efa00d57f407404542370df8547ef
